xla/hlo/transforms/collectives/rematerialize_large_all_gather.cc

```diff
@@ -0,0 +1,165 @@
+   */
+  bool tp_replica_group_found = false;
+  std::vector<ReplicaGroup> tp_replica_group;
+  for (HloInstruction* instruction : computation->MakeInstructionPostOrder()) {
+    if (Match(instruction,
+              match::ReduceScatter().WithOperand(
+                  0, match::AnyOf<HloInstruction>(
+                         match::Op().WithOpcode(HloOpcode::kDot),
+                         match::Op()
+                             .WithOpcode(HloOpcode::kReshape)
+                             .WithOperand(0, match::Op().WithOpcode(
+                                                 HloOpcode::kDot)))))) {
+      tp_replica_group = instruction->replica_groups();
+      tp_replica_group_found = true;
+    }
+    if (tp_replica_group_found) break;
+  }
+  return std::make_pair(tp_replica_group_found, tp_replica_group);
+}
+
+std::pair<bool, std::vector<ReplicaGroup>>
+RematerializeLargeAllGather::GetTpReplicaGroup(
+    HloModule* module,
+    const absl::flat_hash_set<absl::string_view>& execution_threads) {
+  std::vector<ReplicaGroup> tp_replica_group;
+  bool tp_replica_group_found = false;
+  for (HloComputation* computation : module->computations(execution_threads)) {
+    auto rval = GetTpReplicaGroup(computation);
+    if (rval.first) {
+      tp_replica_group = rval.second;
+      tp_replica_group_found = true;
+      break;
+    }
+  }
+  return std::make_pair(tp_replica_group_found, tp_replica_group);
+}
+
+bool RematerializeLargeAllGather::IsRemattableAllGather(
+    std::vector<ReplicaGroup> tp_replica_group, HloInstruction* input_inst) {
+  if (disable_pattern_match_ && !tp_replica_group.empty()) {
+    VLOG(2) << "Matched all-gather with replica group "
+            << input_inst->ToString();
+    return hlo_query::HasMatchingReplicaGroups(input_inst, tp_replica_group);
+  } else {
+    int64_t ag_numel = 1;
+    for (int64_t dim : input_inst->shape().dimensions()) {
+      ag_numel *= dim;
+    }
+    const int64_t ag_size = ag_numel * ShapeUtil::ByteSizeOfPrimitiveType(
+                                           input_inst->shape().element_type());
+    VLOG(2) << "Matched all-gather based on size " << ag_size;
+    return ag_size >= remat_size_in_bytes_;
+  }
+}
+
+StatusOr<bool> RematerializeLargeAllGather::Run(
+    HloModule* module,
+    const absl::flat_hash_set<absl::string_view>& execution_threads) {
+  bool changed = false;
+  bool tp_replica_group_found = false;
+  std::vector<ReplicaGroup> tp_replica_group;
+  if (!disable_pattern_match_) {
+    auto rval = GetTpReplicaGroup(module, execution_threads);
+    if (rval.first) {
+      tp_replica_group_found = true;
+      tp_replica_group = rval.second;
+    }
+  }
+
+  for (HloComputation* computation : module->computations(execution_threads)) {
+    for (HloInstruction* inst : computation->instructions()) {
+      if (inst->opcode() == HloOpcode::kGetTupleElement) {
+        HloInstruction* opt_barrier = inst->mutable_operand(0);
```

Comment:
Outline this into a function? I think we also have helper functions to iterate over all instructions in a module if you find that useful, see xla/service/gpu/hlo_traversal.h

Reply:
Done